include/boost/random/detail/qrng_base.hpp

```diff
@@ -0,0 +1,254 @@
+  //!(s = X::dimension()) vector at each invocation. When all elements are
+  //!exhausted, X::operator() begins anew with the starting element of a
+  //!subsequent s-dimensional vector.
+  //!
+  //!Throws: range_error.
+  result_type operator()()
+  {
+    return curr_elem != dimension() ? load_cached(): next_state();
+  }
+
+  //!Fills a range with quasi-random values.
+  template<typename Iter> void generate(Iter first, Iter last)
+  {
+    for (; first != last; ++first)
+      *first = this->operator()();
+  }
+
+  //!Effects: Advances *this state as if z consecutive
+  //!X::operator() invocations were executed.
+  //!
+  //!Throws: range_error.
+  void discard(boost::uintmax_t z)
+  {
+    const std::size_t dimension_value = dimension();
+
+    // Compiler knows how to optimize subsequent x / y and x % y
+    // statements. In fact, gcc does this even at -O1, so don't
+    // be tempted to "optimize" % via subtraction and multiplication.
+
+    boost::uintmax_t vec_n = z / dimension_value;
+    std::size_t carry = curr_elem + (z % dimension_value);
+
+    vec_n += carry / dimension_value;
+    carry  = carry % dimension_value;
+
+    // Avoid overdiscarding by branchlessly correcting the triple
+    // (D, S + 1, 0) to (D, S, D) (see equality operator)
+    const bool corr = (!carry) & static_cast<bool>(vec_n);
+
+    // Discards vec_n (with correction) consecutive s-dimensional vectors
+    discard_vector(vec_n - static_cast<boost::uintmax_t>(corr));
+    // Sets up the proper position of the element-to-read
+    // curr_elem = carry + corr*dimension_value
+    curr_elem = carry ^ (-static_cast<std::size_t>(corr) & dimension_value);
+  }
+
+  //!Writes the textual representation of the generator to a @c std::ostream.
+  BOOST_RANDOM_DETAIL_OSTREAM_OPERATOR(os, qrng_base, s)
+  {
+    os << s.dimension() << " " << s.seq_count << " " << s.curr_elem;
+    return os;
+  }
+
+  //!Reads the textual representation of the generator from a @c std::istream.
+  BOOST_RANDOM_DETAIL_ISTREAM_OPERATOR(is, qrng_base, s)
+  {
+    std::size_t dim;
+    size_type seed;
+    boost::uintmax_t z;
+    if (is >> dim >> std::ws >> seed >> std::ws >> z) // initialize iff success!
+    {
+      if (s.dimension() != prevent_zero_dimension(dim))
+      {
+        s.lattice.resize(dim);
+        s.quasi_state.resize(dim);
+      }
+      // Fast-forward to the correct state
+      s.derived().seed(seed);
+      if (z != 0) s.discard(z);
+    }
+    return is;
+  }
+
+  //!Returns true if the two generators will produce identical sequences of outputs.
+  BOOST_RANDOM_DETAIL_EQUALITY_OPERATOR(qrng_base, x, y)
+  {
+    const std::size_t dimension_value = x.dimension();
+
+    // Note that two generators with different seq_counts and curr_elems can
+    // produce the same sequence because the generator triple
+    // (D, S, D) is equivalent to (D, S + 1, 0), where D is dimension, S -- seq_count,
+    // and the last one is curr_elem.
```

Comment:
Would it make sense to normalize the representation?

Reply:
It is normalized. The difference between (D, S, D) and (D, S + 1, 0) is that the former has all the elements exhausted but the new state is not yet computed (because not yet requested), while the latter is a freshly computed S + 1 state (because it was requested).

It is useful to maintain this difference, because otherwise if you do engine.discard(maxdiscard) and compute straight to state (D, last_state + 1, 0), you would be overdiscarding, you wouldn't have the equality between maxdiscard operator() invocations and discard(maxdiscard).